### PR TITLE
messagestore implementation

### PR DESCRIPTION
**Pull request by Mathiesen**
## Description

Implement MessageStore 

## References

* #0000

### CODE REVIEW

File: source/B2B.Transactions.CimMessageAdapter/Errors/DuplicateMessageIdDetected.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ namespace B2B.CimMessageAdapter.Errors
 {
     public class DuplicateMessageIdDetected : ValidationError
     {
-        public DuplicateMessageIdDetected(string messageId)
+        public DuplicateMessageIdDetected(string? messageId)
             : base($"Message id '{messageId}' is not unique", "B2B-003")
```
Comment (kft) [inline source/B2B.Transactions.CimMessageAdapter/Errors/DuplicateMessageIdDetected.cs:20]: Message id can never be null. Why this change ?

File: source/B2B.Transactions.CimMessageAdapter/Errors/DuplicateTransactionIdDetected.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ namespace B2B.CimMessageAdapter.Errors
 {
     public class DuplicateTransactionIdDetected : ValidationError
     {
-        public DuplicateTransactionIdDetected(string transactionId)
+        public DuplicateTransactionIdDetected(string? transactionId)
             : base($"Transaction id '{transactionId}' is not unique and will not be processed.", "B2B-005")
```
Comment (kft) [inline source/B2B.Transactions.CimMessageAdapter/Errors/DuplicateTransactionIdDetected.cs:20]: Same as above

File: source/B2B.Transactions.CimMessageAdapter/Errors/UnknownReceiver.cs [hunk 3]
```diff
@@ -16,7 +16,7 @@ namespace B2B.CimMessageAdapter.Errors
 {
     public class UnknownReceiver : ValidationError
     {
-        public UnknownReceiver(string receiverId)
+        public UnknownReceiver(string? receiverId)
             : base($"Receiver id {receiverId} is not known receiver", "B2B-008")
```
Comment (kft) [inline source/B2B.Transactions.CimMessageAdapter/Errors/UnknownReceiver.cs:20]: Same as above

File: source/B2B.Transactions.CimMessageAdapter/Messages/IMessageIds.cs [hunk 4]
```diff
@@ -25,6 +25,6 @@ public interface IMessageIds
         /// Store message id
         /// </summary>
         /// <param name="messageId"></param>
-        Task<bool> TryStoreAsync(string messageId);
+        Task<bool> TryStoreAsync(string? messageId);
     }
```
Comment (kft) [inline source/B2B.Transactions.CimMessageAdapter/Messages/IMessageIds.cs:29]: Same as above

File: source/B2B.Transactions.CimMessageAdapter/Messages/ITransactionIds.cs [hunk 5]
```diff
@@ -25,6 +25,6 @@ public interface ITransactionIds
         /// Store transaction id
         /// </summary>
         /// <param name="transactionId"></param>
-        Task<bool> TryStoreAsync(string transactionId);
+        Task<bool> TryStoreAsync(string? transactionId);
```
Comment (kft) [inline source/B2B.Transactions.CimMessageAdapter/Messages/ITransactionIds.cs:28]: Same as above

File: source/B2B.Transactions.Infrastructure/DataAccess/Transaction/TransactionRepository.cs [hunk 6]
```diff
@@ -31,7 +31,7 @@ public void Add(AcceptedTransaction acceptedTransaction)
             _b2BContext.Transactions.Add(acceptedTransaction);
         }
 
-        public AcceptedTransaction? GetById(string transactionId)
+        public AcceptedTransaction? GetById(string? transactionId)
         {
```
Comment (kft) [inline source/B2B.Transactions.Infrastructure/DataAccess/Transaction/TransactionRepository.cs:35]: Transaction id can never be null here.

File: source/B2B.Transactions.Infrastructure/Messages/MessageIdRegistry.cs [hunk 7]
```diff
@@ -29,7 +29,7 @@ public MessageIdRegistry(IDbConnectionFactory connectionFactory)
             _connectionFactory = connectionFactory ?? throw new ArgumentNullException(nameof(connectionFactory));
         }
 
-        public async Task<bool> TryStoreAsync(string messageId)
+        public async Task<bool> TryStoreAsync(string? messageId)
         {
```
Comment (kft) [inline source/B2B.Transactions.Infrastructure/Messages/MessageIdRegistry.cs:33]: Same as above

File: source/B2B.Transactions.Infrastructure/Messages/TransactionIdRegistry.cs [hunk 8]
```diff
@@ -29,7 +29,7 @@ public TransactionIdRegistry(IDbConnectionFactory connectionFactory)
             _connectionFactory = connectionFactory ?? throw new ArgumentNullException(nameof(connectionFactory));
         }
 
-        public async Task<bool> TryStoreAsync(string transactionId)
+        public async Task<bool> TryStoreAsync(string? transactionId)
         {
```
Comment (kft) [inline source/B2B.Transactions.Infrastructure/Messages/TransactionIdRegistry.cs:33]: Same as above

File: source/B2B.Transactions.Infrastructure/OutgoingMessageStore.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                var unpublished = _context
+                    .OutgoingMessages
+                    .Where(x => x.IsPublished == false)
+                    .ToList();
+
+                return unpublished;
+            }).ConfigureAwait(false);
+
+            return retVal.Count == 0 ? new ReadOnlyCollection<OutgoingMessage>(new List<OutgoingMessage>())
+                : new ReadOnlyCollection<OutgoingMessage>(retVal);
```
Comment (kft) [inline source/B2B.Transactions.Infrastructure/OutgoingMessageStore.cs:51]: Why are we starting a new task here ? Why not just query the DbSet directly ?

File: source/B2B.Transactions.Infrastructure/OutgoingMessages/MessagePublisher.cs [hunk 10]
```diff
@@ -51,7 +51,7 @@ private static DataAvailableNotificationDto CreateDataAvailableNotificationFrom(
         {
             return new DataAvailableNotificationDto(
                 Guid.NewGuid(),
-                new GlobalLocationNumberDto(message.RecipientId),
+                new GlobalLocationNumberDto(message.RecipientId ?? string.Empty),
                 new MessageTypeDto(string.Empty),
```
Comment (kft) [inline source/B2B.Transactions.Infrastructure/OutgoingMessages/MessagePublisher.cs:55]: Recipient id should never be null.

File: source/B2B.Transactions.IntegrationTests/Infrastructure/OutgoingMessages/MessagePublisherTests.cs [hunk 11]
```diff
@@ -30,7 +30,7 @@ namespace B2B.Transactions.IntegrationTests.Infrastructure.OutgoingMessages
     public class MessagePublisherTests : TestBase
     {
         private readonly IOutgoingMessageStore _outgoingMessageStore;
-        private readonly IMessageFactory<IDocument> _messageFactory;
+        private readonly IMessageFactory<IDocument?> _messageFactory;
         private readonly MessagePublisher _messagePublisher;
```
Comment (kft) [inline source/B2B.Transactions.IntegrationTests/Infrastructure/OutgoingMessages/MessagePublisherTests.cs:34]: Null here ?

File: source/B2B.Transactions.IntegrationTests/OutgoingMessageStoreTests.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            _testHelper = new TestHelper(
+                GetService<ITransactionRepository>(),
+                GetService<IUnitOfWork>(),
+                GetService<IOutgoingMessageStore>(),
+                GetService<IMessageFactory<IDocument>>());
+        }
+
+        [Fact]
+        public async Task Can_get_unpublished_message_from_messagestore()
+        {
```
Comment (kft) [inline source/B2B.Transactions.IntegrationTests/OutgoingMessageStoreTests.cs:43]: This test class not be needed at all. Outgoing message store is an implemenation details, and is test indirectly in the MessagePublisherTests

File: source/B2B.Transactions.IntegrationTests/TestHelper.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            IOutgoingMessageStore messageStore,
+            IMessageFactory<IDocument?> messageFactory)
+        {
+            _transactionRepository = transactionRepository;
+            UnitOfWork = unitOfWork;
+            MessageStore = messageStore;
+            _messageFactory = messageFactory;
+        }
+
+        public IOutgoingMessageStore MessageStore { get; }
```
Comment (kft) [inline source/B2B.Transactions.IntegrationTests/TestHelper.cs:42]: There is already a TransactionBuilder in the test project. No need for a TestHelper

File: source/B2B.Transactions.IntegrationTests/Transactions/TransactionHandlingTests.cs [hunk 14]
```diff
@@ -32,7 +32,7 @@ public class TransactionHandlingTests : TestBase
         private readonly IUnitOfWork _unitOfWork;
         private readonly XNamespace _namespace = "urn:ediel.org:structure:confirmrequestchangeofsupplier:0:1";
         private OutgoingMessageStoreSpy _outgoingMessageStoreSpy = new();
-        private IMessageFactory<IDocument> _messageFactory = new AcceptMessageFactory(_dateTimeProvider);
+        private IMessageFactory<IDocument?> _messageFactory = new AcceptMessageFactory(_dateTimeProvider);
```
Comment (kft) [inline source/B2B.Transactions.IntegrationTests/Transactions/TransactionHandlingTests.cs:36]: Nullable ?

File: source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs [hunk 15]
```diff
@@ -21,8 +21,8 @@ public AcceptMessage(string messagePayload, string messageType)
             DocumentType = messageType;
         }
 
-        public string MessagePayload { get; init; }
+        public string MessagePayload { get; set; }
```
Comment (kft) [inline source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs:25]: MessagePayload should not be changed after creation.

File: source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs [hunk 16]
```diff
@@ -21,8 +21,8 @@ public AcceptMessage(string messagePayload, string messageType)
...
 
-        public string DocumentType { get; init; }
+        public string DocumentType { get; set; }
     }
```
Comment (kft) [inline source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs:27]: Same as above

File: source/B2B.Transactions/OutgoingMessages/IDocument.cs [hunk 17]
```diff
@@ -22,11 +22,11 @@ public interface IDocument
         /// <summary>
         /// Payload of the outgoing message
         /// </summary>
-        string MessagePayload { get; init; }
+        string MessagePayload { get; set; }
```
Comment (kft) [inline source/B2B.Transactions/OutgoingMessages/IDocument.cs:25]: Same as above

File: source/B2B.Transactions/OutgoingMessages/IDocument.cs [hunk 18]
```diff
@@ -22,11 +22,11 @@ public interface IDocument
...
 
         /// <summary>
         /// The type of the message
         /// </summary>
-        string DocumentType { get; init; }
+        string DocumentType { get; set; }
```
Comment (kft) [inline source/B2B.Transactions/OutgoingMessages/IDocument.cs:30]: Same as above

File: source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs [hunk 19]
```diff
@@ -18,19 +18,40 @@ namespace B2B.Transactions.OutgoingMessages
 {
     public class OutgoingMessage
     {
-        public OutgoingMessage(IDocument document, string recipientId)
+        public OutgoingMessage(IDocument? document, string? recipientId)
         {
```
Comment (kft) [inline source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs:22]: None of these can be null.

File: source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs [hunk 20]
```diff
@@ -18,19 +18,40 @@ namespace B2B.Transactions.OutgoingMessages
...
             RecipientId = recipientId;
         }
 
-        public IDocument Document { get; }
+        public OutgoingMessage()
+        { }
+
+        public Guid Id { get; }
+
+        public IDocument? Document { get; }
```
Comment (kft) [inline source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs:33]: Can't be null

File: source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs [hunk 21]
```diff
@@ -18,19 +18,40 @@ namespace B2B.Transactions.OutgoingMessages
...
+        }
+
+        public string MessagePayload
+        {
+            get => Document?.MessagePayload ?? string.Empty;
+            set
+            {
+                if (Document != null) Document.MessagePayload = value;
+            }
+        }
```
Comment (kft) [inline source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs:54]: Same as above

File: source/B2B.Transactions/Transactions/AcceptedTransaction.cs [hunk 22]
```diff
@@ -16,11 +16,11 @@ namespace B2B.Transactions.Transactions
 {
     public class AcceptedTransaction
     {
-        public AcceptedTransaction(string transactionId)
+        public AcceptedTransaction(string? transactionId)
```
Comment (kft) [inline source/B2B.Transactions/Transactions/AcceptedTransaction.cs:19]: Same as above

File: source/B2B.Transactions/Transactions/AcceptedTransaction.cs [hunk 23]
```diff
@@ -16,11 +16,11 @@ namespace B2B.Transactions.Transactions
...
         {
             TransactionId = transactionId;
         }
 
-        public string TransactionId { get; }
+        public string? TransactionId { get; }
```
Comment (kft) [inline source/B2B.Transactions/Transactions/AcceptedTransaction.cs:24]: Same as above

File: source/B2B.Transactions/Transactions/ITransactionRepository.cs [hunk 24]
```diff
@@ -32,6 +32,6 @@ public interface ITransactionRepository
         /// </summary>
         /// <param name="transactionId"></param>
         /// <returns><see cref="AcceptedTransaction"/></returns>
-        AcceptedTransaction? GetById(string transactionId);
+        AcceptedTransaction? GetById(string? transactionId);
```
Comment (kft) [inline source/B2B.Transactions/Transactions/ITransactionRepository.cs:35]: Same as above

**Review by kft — CHANGES_REQUESTED** [review]

File: source/B2B.Transactions.CimMessageAdapter/Errors/DuplicateTransactionIdDetected.cs (hunk 2 above)
Comment (CBrohus) [reply source/B2B.Transactions.CimMessageAdapter/Errors/DuplicateTransactionIdDetected.cs]: Changed to non-nullable

File: source/B2B.Transactions.CimMessageAdapter/Errors/UnknownReceiver.cs (hunk 3 above)
Comment (CBrohus) [reply source/B2B.Transactions.CimMessageAdapter/Errors/UnknownReceiver.cs]: Changed to non-nullable

File: source/B2B.Transactions.CimMessageAdapter/Messages/IMessageIds.cs (hunk 4 above)
Comment (CBrohus) [reply source/B2B.Transactions.CimMessageAdapter/Messages/IMessageIds.cs]: Changed to non-nullable

File: source/B2B.Transactions.CimMessageAdapter/Messages/ITransactionIds.cs (hunk 5 above)
Comment (CBrohus) [reply source/B2B.Transactions.CimMessageAdapter/Messages/ITransactionIds.cs]: Changed to non-nullable

File: source/B2B.Transactions.Infrastructure/Messages/MessageIdRegistry.cs (hunk 7 above)
Comment (CBrohus) [reply source/B2B.Transactions.Infrastructure/Messages/MessageIdRegistry.cs]: Changed to non-nullable

File: source/B2B.Transactions.Infrastructure/Messages/TransactionIdRegistry.cs (hunk 8 above)
Comment (CBrohus) [reply source/B2B.Transactions.Infrastructure/Messages/TransactionIdRegistry.cs]: Changed to non-nullable

File: source/B2B.Transactions.Infrastructure/OutgoingMessageStore.cs (hunk 9 above)
Comment (CBrohus) [reply source/B2B.Transactions.Infrastructure/OutgoingMessageStore.cs]: Changed to query DBSet directly. Superfluous Task removed

File: source/B2B.Transactions.Infrastructure/OutgoingMessages/MessagePublisher.cs (hunk 10 above)
Comment (CBrohus) [reply source/B2B.Transactions.Infrastructure/OutgoingMessages/MessagePublisher.cs]: Changed to non-nullable

File: source/B2B.Transactions.IntegrationTests/Infrastructure/OutgoingMessages/MessagePublisherTests.cs (hunk 11 above)
Comment (CBrohus) [reply source/B2B.Transactions.IntegrationTests/Infrastructure/OutgoingMessages/MessagePublisherTests.cs]: Changed to non-nullable

File: source/B2B.Transactions.IntegrationTests/Transactions/TransactionHandlingTests.cs (hunk 14 above)
Comment (CBrohus) [reply source/B2B.Transactions.IntegrationTests/Transactions/TransactionHandlingTests.cs]: Changed to non-nullable

File: source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs (hunk 15 above)
Comment (CBrohus) [reply source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs]: Changed to "init" to ensure MessagePayload can only be set once

File: source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs (hunk 16 above)
Comment (CBrohus) [reply source/B2B.Transactions/OutgoingMessages/AcceptMessage.cs]: Changed to "init" to ensure DocumentType can only be set once

File: source/B2B.Transactions/OutgoingMessages/IDocument.cs (hunk 17 above)
Comment (CBrohus) [reply source/B2B.Transactions/OutgoingMessages/IDocument.cs]: Changed to "init" to ensure MessagePayload can only be set once

File: source/B2B.Transactions/OutgoingMessages/IDocument.cs (hunk 18 above)
Comment (CBrohus) [reply source/B2B.Transactions/OutgoingMessages/IDocument.cs]: Changed to "init" to ensure DocumentType can only be set once

File: source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs (hunk 19 above)
Comment (CBrohus) [reply source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs]: Changed to non-nullable

File: source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs (hunk 20 above)
Comment (CBrohus) [reply source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs]: Changed to non-nullable

File: source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs (hunk 21 above)
Comment (CBrohus) [reply source/B2B.Transactions/OutgoingMessages/OutgoingMessage.cs]: Changed to non-nullable

File: source/B2B.Transactions/Transactions/AcceptedTransaction.cs (hunk 22 above)
Comment (CBrohus) [reply source/B2B.Transactions/Transactions/AcceptedTransaction.cs]: Changed to non-nullable

File: source/B2B.Transactions/Transactions/AcceptedTransaction.cs (hunk 23 above)
Comment (CBrohus) [reply source/B2B.Transactions/Transactions/AcceptedTransaction.cs]: Changed to non-nullable

File: source/B2B.Transactions/Transactions/ITransactionRepository.cs (hunk 24 above)
Comment (CBrohus) [reply source/B2B.Transactions/Transactions/ITransactionRepository.cs]: Changed to non-nullable

File: source/B2B.Transactions.IntegrationTests/TestHelper.cs (hunk 13 above)
Comment (CBrohus) [reply source/B2B.Transactions.IntegrationTests/TestHelper.cs]: Moved necessary code into TransactionBuilder and removed TestHelper.cs

File: source/B2B.Transactions.IntegrationTests/OutgoingMessageStoreTests.cs (hunk 12 above)
Comment (CBrohus) [reply source/B2B.Transactions.IntegrationTests/OutgoingMessageStoreTests.cs]: Correct. The included test is already in MessagePublisherTests.

Deleted OutgoingMessageStoreTests.cs.

**Pull request merged**